packages/flutter_tools/lib/src/custom_devices/custom_devices_config.dart

```diff
@@ -0,0 +1,97 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'package:meta/meta.dart';
+
+import '../base/config.dart';
+import '../base/file_system.dart';
+import '../base/logger.dart';
+import '../base/platform.dart';
+import 'custom_device_config.dart';
+
+/// Represents the custom devices config file on disk which in turn
+/// contains a list of individual custom device configs.
+class CustomDevicesConfig {
+  /// Load a [CustomDevicesConfig] from a (possibly non-existent) location on disk.
+  ///
+  /// The config is loaded on construction. Any error while loading will be logged
+  /// but will not result in an exception being thrown. The file will not be deleted
+  /// when it's not valid JSON (which other configurations do) and will not
+  /// be implicitly created when it doesn't exist.
+  CustomDevicesConfig({
+    @required FileSystem fileSystem,
+    @required Logger logger,
+    @required Platform platform
+  }) : _fileSystem = fileSystem,
+       _config = Config(
+         _kCustomDevicesConfigName,
+         fileSystem: fileSystem,
+         logger: logger,
+         platform: platform,
+         deleteFileOnFormatException: false
+       )
+  {
+    ensureFileExists();
+  }
+
+  @visibleForTesting
+  CustomDevicesConfig.test({
+    @required FileSystem fileSystem,
+    Directory directory,
+    @required Logger logger
+  }) : _fileSystem = fileSystem,
+       _config = Config.test(
+         name: _kCustomDevicesConfigName,
+         directory: directory,
+         logger: logger,
+         deleteFileOnFormatException: false
+       )
+  {
+    ensureFileExists();
+  }
+
+  static const String _kCustomDevicesConfigName = 'custom_devices.json';
+  static const String _kCustomDevicesConfigKey = 'custom-devices';
+  static const String _kSchema = r'$schema';
+  static const String _defaultSchema = 'https://raw.githubusercontent.com/flutter/flutter/stable/packages/flutter_tools/static/custom-devices.schema.json';
```

Comment:
I think this should actually refer to the schema via local path into the SDK instead of from github. Otherwise if we make changes locally or in a pre-release branch (dev/beta) users may see the incorrect configuration until that reaches stable.

You can use Cache.flutterRoot to find the root of the flutter checkout, then path manipulation from there.

Reply:
Yeah I see, a file path seems to be the better compromise.